### ISSUE TITLE
Discourage or deprecate detached configurations

### ISSUE THREAD

**Issue by tadfisher** [gradle/gradle#29489]
### Expected Behavior

Use of `Project.detachedConfiguration` and `Configuration.detachedConfiguration` should be deprecated for future removal, and the alternative should be to register normal configurations on ConfigurationContainer that can be analyzed and manipulated at configuration time.

### Context

Detached configurations (as created with `Project.detachedConfiguration` or `ConfigurationContainer.detachedConfiguration`) are problematic and hostile for development.

- These configurations may be created in the execution phase, bypassing configuration caching and introducing potential impurities to reproducible builds.
- These configurations are not resolved by the `dependencies` task.
- These configurations are missed by the [GitHub Dependency Graph Gradle Plugin](https://github.com/gradle/github-dependency-graph-gradle-plugin/), a tool published under the Gradle org that is meant to help developers create SBOMs and analyze dependencies for security vulnerabilities.
- Overriding dependencies or customizing resolution in these configurations is impossible without forking plugin code or using dependency substitution with included builds, something that isn't needed for "normal" configurations.

Essentially, these configurations do not exist until the code that creates them is executed, usually as part of task creation. Thus they should be discouraged.

**Comment by tadfisher** [comment]
Another problem:
- Dependency locking using `gradle dependencies --write-locks` will fail to lock dependency versions that only appear in detached configurations.

**Comment by tadfisher** [comment]
Another problem:
- Dependency verification with `gradle --write-verification-metadata sha256 help` will fail to record dependency artifacts that are only resolved by detached configurations.

**Comment by hegyibalint** [comment]
This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by big-guy** [comment]
I agree with your points about the problems with detached configurations.  These are being used in a lot of important plugins and we need to provide alternatives for those plugins to migrate to that do not have the same problems as detached configurations.

We won't be able to deprecate detached configurations soon (in 8.x), but we can make progress on supporting different use cases through ArtifactViews/Configuration or new APIs.

related #9894

**Comment by tadfisher** [comment]
Is there a technical issue with the proposed alternative of regular configurations? Potential naming clash, perhaps?

**Comment by Vampire** [comment]
Imho not only potential naming clash although that is one point.
There are also other points.
For example using regular configurations for plugin internal constructs clutters the namespace that is available in the build script and the consumer does not know what he should or should not configure.
Also if you expose such plugin internals, you risk that the consumer modifies them which he maybe should not be able to.

That those configurations are not resolved by the `dependencies` task and not handled by the GitHub dependency graph plugin, are also a good thing. They are typically no dependencies of the project in question, but implementation details of the plugins the build for that project uses. They should not be treated like dependencies of the project.

If a plugin wants the consumer to be able to modify the dependencies, it should expose the configurations like the built-in code quality plugins do it. But there are way more use-cases where detached configurations are useful and that should not be forced to be exposed imho.

**Comment by tadfisher** [comment]
I want to address @Vampire's point here:

> That those configurations are not resolved by the dependencies task and not handled by the GitHub dependency graph plugin, are also a good thing. They are typically no dependencies of the project in question, but implementation details of the plugins the build for that project uses. They should not be treated like dependencies of the project.

This is not viable if the goal is to prevent supply chain attacks. If we cannot verify plugin dependencies, even if they are an implementation detail, we cannot prevent malicious substitutions.

Reproducible builds are more difficult to achieve as well. We need the entire dependency closure, including the dependencies of detached configurations, in order to reproduce the build. Today this can be done with hacks; for example, see the many uses of the `implicitDependencies` configuration in the [kotlin repo](https://github.com/search?q=repo%3AJetBrains%2Fkotlin%20implicitDependencies&type=code).

**Comment by Vampire** [comment]
Sure, but you would also need to prevent any other possibility for a plugin to download something.
A plugin can also leverage the undercouch download plugin, or just use the built-in HTTP client, or even an URL connection to download things it needs.
Detached configurations are just another way to download something with the added value of Gradle doing dependency resolution before. :-)

**Comment by Vampire** [comment]
I don't know what you mean.
If the plugin using that detached configuration is using strict versions or dependency constraints, it should just work fine.
If you mean from outside the plugin then I still don't know what you mean
How would you configure any of it, as you don't get a hold on the detached configurations, they are detached.

Again, consider a detached configuration like the plugin using HttpClient or UrlConnection or OkHttp or ktor-client or any other means of downloading a file. It is a plugin internal implementation detail that you as plugin user do not have any influence on, unless the plugin exposes such influence possibilities to its consumers.

**Comment by chrisjenx** [comment]
That would make sense if those detached configurations didn't use the repositories we define and didn't use the gradle dependency chain to download them, the issue is that they do, but ignore dependency constraints, but will blow up when it can't find those dependencies, despite overriding them with force/constraints.

Which is it, are they external clients, or do they use Gradle dependency system?

**Comment by nskvortsov** [comment]
The detached configurations are perfectly viable instrument for the use cases described by @Vampire above. Plugins need to obtain their own tools, compilers, libraries and so on.  None of these dependencies should land in project dependencies, obviously.

Obviously, regular dependencies reports will miss detached configurations.
Still, they should be optionally reported after the build execution. 

Also, they could be subjects of global dependencies constraints or checksum verification in runtime.
